src/nvim/drawline.c

```diff
@@ -1723,7 +1749,87 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
         area_active = false;
       }
 
-      if (!wlv.n_extra) {
+      if (has_decor && v >= 0) {
```

Comment:
Can this whole block be factored into a function. `win_line()` is bad enough as it is.